File: rails/app/models/concern.rb
class Concern < ApplicationRecord
  before_validation { self.trigger_event = "" if trigger_event.nil? }

  validates :content, presence: true, length: { maximum: 1000 }
  validates :trigger_event, length: { maximum: 120 }

  belongs_to :user
  has_one :issue, dependent: :destroy
  has_one :roadmap, dependent: :destroy

  scope :active, -> { where(archived_at: nil) }
  scope :archived, -> { where.not(archived_at: nil) }

  AUTO_ARCHIVE_PERIOD = 7.days

  def archive!
    return if archived?

    update!(
      archived_at: Time.current,
      auto_archive_at: nil,
    )
  end

  def unarchive!
    return unless archived?

    update!(
      archived_at: nil,
      auto_archive_at: user.auto_archive_enabled? ? Time.current + AUTO_ARCHIVE_PERIOD : nil,
    )
  end

  def archived?
    archived_at.present?
  end
end

File: next/src/types/concern.ts
import { Issue, IssueResponse } from "./issue";
import { Roadmap, RoadmapResponse } from "./roadmap";

export type ConcernResponse = {
  id: number;
  trigger_event: string;
  content: string;
  archived_at: string | null;
  created_at: string;
};

export type ConcernDetailResponse = {
  concern: ConcernResponse;
  issue: IssueResponse | null;
  roadmap: RoadmapResponse | null;
};

export type Concern = {
  id: number;
  triggerEvent: string;
  content: string;
  archivedAt: string | null;
  createdAt: string;
};

export type ConcernInput = {
  triggerEvent: string;
  content: string;
};

export type ConcernDetail = {
  concern: Concern;
  issue: Issue | null;
  roadmap: Roadmap | null;
};
